### PR TITLE
Implement Azure Blob Export Sink

### PR DESCRIPTION
**Pull request by wsugarman**
## Description
- Implement `AzureBlobExportSink`, and its corresponding providers
- Add `ExportService`
- Add new registration methods
- Refactor operations client:
  - Accept operation Id as input instead of generating it. This way callers can perform actions based on the ID of the job they're about to submit (e.g. for export, the store uploads a secret with the job ID as the key)
  - Standard APIs to return an `OperationReference` for the operation they submitted
- Remove `ISensitive` as it's no longer needed

## Related issues
Addresses [AB#90232](https://microsofthealth.visualstudio.com/Health/_workitems/edit/90232)
Addresses [AB#91160](https://microsofthealth.visualstudio.com/Health/_workitems/edit/91160)

## Testing
New Tests

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Core/Features/Common/GuidFactory.cs [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 using System;
 
-namespace Microsoft.Health.Dicom.Functions.Client.DurableTask;
+namespace Microsoft.Health.Dicom.Core.Features.Common;
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Core/Features/Common/GuidFactory.cs:8]: I am curious, why do we need a factory to get the functionality of Guid.NewGuid()?

Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Common/GuidFactory.cs:8]: Mostly so I can inject a deterministic `Guid` value during testing, like `Clock` for `DateTime.Now`

File: src/Microsoft.Health.Dicom.Blob/Features/Export/ExportPatternPlaceholder.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+
+[Flags]
+internal enum ExportPatternPlaceholders : sbyte
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Blob/Features/Export/ExportPatternPlaceholder.cs:9]: Do not understand. Do we need this? Should we need to parameterize this?

File: src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+using Microsoft.Health.Dicom.Core.Features.Common;
+using Microsoft.Health.Dicom.Core.Features.Export;
+using Microsoft.Health.Dicom.Core.Models.Common;
+using Microsoft.Health.Dicom.Core.Models.Export;
+
+namespace Microsoft.Health.Dicom.Blob.Features.Export;
+
+internal sealed class AzureBlobExportSink : IExportSink
+{
+    public Uri ErrorHref
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs:31]: nit: should this go after the constructor?

File: src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    private readonly AzureBlobExportFormatOptions _output;
+    private readonly BlobOperationOptions _blobOptions;
+    private readonly JsonSerializerOptions _jsonOptions;
+
+    private const int BlockSize = 2 * 1024 * 1024;
+
+    public AzureBlobExportSink(
+        IFileStore source,
+        BlobContainerClient dest,
+        IOptions<AzureBlobExportFormatOptions> outputOptions,
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs:58]: If this is not user controlled, should this be passed in or implicit?

File: src/Microsoft.Health.Dicom.Blob/Features/Export/ExportFilePattern.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+using System.Text;
+using EnsureThat;
+using Microsoft.Health.Dicom.Core.Features.Model;
+using Microsoft.Health.Operations;
+
+namespace Microsoft.Health.Dicom.Blob.Features.Export;
+
+internal static class ExportFilePattern
+{
+    public static string Parse(string pattern, ExportPatternPlaceholders placeholders = ExportPatternPlaceholders.All)
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Blob/Features/Export/ExportFilePattern.cs:17]: Yeah, I can add some comments. Here I am trying to parse from some format of like `/%Study%/%Series%/%SopInstance%.dcm` to a .NET format string like `{0}/{1}/{2}.dcm`. The idea was to provide users with a convenient way of specifying their custom paths without having to use our own format strings with numbers. However, I needed some translation layer in our code

File: src/Microsoft.Health.Dicom.Core/Features/Export/ExportService.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+        // Validate the input and update the specification
+        specification = new ExportSpecification
+        {
+            Source = await _sourceFactory.ValidateAsync(specification.Source, cancellationToken),
+            Destination = await _sinkFactory.ValidateAsync(specification.Destination, operationId, cancellationToken),
+        };
+
+        // Start the operation
+        return await _client.StartExportAsync(operationId, specification, _partition, cancellationToken);
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Core/Features/Export/ExportService.cs:55]: I was expecting SAS token to be put in the KV here? Did I miss that code? I also did see it being read in Export Sink

File: src/Microsoft.Health.Dicom.Blob/Features/Export/ExportFilePattern.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (smithago) [reply src/Microsoft.Health.Dicom.Blob/Features/Export/ExportFilePattern.cs:17]: Lets add it when we expose that feature.

File: src/Microsoft.Health.Dicom.Core/Features/Export/ExportService.cs (hunk 6 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Export/ExportService.cs:55]: As discussed offline, I have separated the logic outside of `ValidateAsync`

File: src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs (hunk 4 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs:58]: I have made the output options `internal`, but we can still pass it in as an options as if it's customizable, should we ever need to change this.

File: src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs (hunk 3 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportSink.cs:31]: Good catch

File: src/Microsoft.Health.Dicom.Blob/Features/Export/ExportPatternPlaceholder.cs (hunk 2 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Blob/Features/Export/ExportPatternPlaceholder.cs:9]: As discussed offline, I went all in on the path specification lol. I have made this `internal` for now

File: src/Microsoft.Health.Dicom.Blob/Features/Export/ExportFilePattern.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Blob/Features/Export/ExportFilePattern.cs:17]: As discussed, I have made this `internal` for now

File: src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportFormatOptions.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Text;
+using EnsureThat;
+using Microsoft.Health.Dicom.Core.Features.Model;
+
+namespace Microsoft.Health.Dicom.Blob.Features.Export;
+
+internal sealed class AzureBlobExportFormatOptions
```
[the remark this answers is not in the document]
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Blob/Features/Export/AzureBlobExportFormatOptions.cs:13]: For now it's an option, but the user can't control it. It's more about for the future if we ever need to change things

File: src/Microsoft.Health.Dicom.Core/Features/Common/GuidFactory.cs (hunk 1 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Features/Common/GuidFactory.cs:8]: Will leave a comment

**Review by smithago — APPROVED** [review]

**Pull request merged**